piq/vsi.py

```diff
@@ -0,0 +1,285 @@
+        r"""Computation of VSI as a loss function.
+
+            Args:
+                prediction: Tensor of prediction of the network.
+                target: Reference tensor.
+
+            Returns:
+                Value of VSI loss to be minimized. 0 <= VSI loss <= 1.
+        """
+
+        return 1. - vsi(prediction=prediction, target=target, reduction=self.reduction, data_range=self.data_range,
+                        c1=self.c1, c2=self.c2, c3=self.c3, alpha=self.alpha, beta=self.beta,
+                        omega_0=self.omega_0, sigma_f=self.sigma_f, sigma_d=self.sigma_d, sigma_c=self.sigma_c)
+
+
+def scharr_grad_map(x: torch.Tensor) -> torch.Tensor:
+    # Schaee filters are used to get gradient maps
+    filter_x = torch.tensor([[3., 0., -3.],
+                             [10., 0., -10.],
+                             [3., 0., -3.]]).to(x) / 16
+
+    filter_y = torch.tensor([[3., 10., 3.],
+                             [0., 0., 0.],
+                             [-3., -10., -3.]]).to(x) / 16
+
+    padding = filter_x.size(-1) // 2
+
+    gm_x = conv2d(x, filter_x.view(1, 1, *filter_x.size()), padding=padding)
+    gm_y = conv2d(x, filter_y.view(1, 1, *filter_y.size()), padding=padding)
+
+    return torch.sqrt(gm_x.pow(2) + gm_y.pow(2))
+
+
+def sdsp(x: torch.Tensor, data_range: float = 255., omega_0: float = 0.021, sigma_f: float = 1.34,
+         sigma_d: float = 145., sigma_c: float = 0.001) -> torch.Tensor:
+    size = x.size()
+
+    size_to_use = (256, 256)
+    x = interpolate(input=x, size=size_to_use, mode='bilinear', align_corners=True)
+    x_lab = rgb2lab(x, data_range=data_range)
+
+    x_fft = torch.fft(torch.stack([x_lab, torch.zeros_like(x_lab)], dim=-1), 2)
+    lg = log_gabor(size_to_use, omega_0, sigma_f).to(x_fft).view(1, 1, *size_to_use, 1)
+    x_ifft_real = torch.ifft(x_fft * lg, 2)[..., 0]
+    s_f = x_ifft_real.pow(2).sum(dim=1).sqrt()
+
+    coordinates = torch.stack((torch.meshgrid([torch.arange(0, size_to_use[0]) - (size[0] - 1) / 2,
+                                               torch.arange(0, size_to_use[1]) - (size[1] - 1) / 2])), dim=0).to(x)
+    s_d = torch.exp(-torch.sum(coordinates, dim=0) / sigma_d ** 2)
+
+    eps = torch.finfo(x_lab.dtype).eps
+    min_x = x_lab.min(dim=-1, keepdim=True).values.min(dim=-2, keepdim=True).values
+    max_x = x_lab.max(dim=-1, keepdim=True).values.max(dim=-2, keepdim=True).values
+    normilized = (x_lab - min_x) / (max_x - min_x + eps)
+
+    norm = normilized[:, 1:].pow(2).sum(dim=1)
+    s_c = 1 - torch.exp(-norm / sigma_c ** 2)
+
+    vs_m = interpolate((s_f * s_d * s_c).unsqueeze(1), size[-2:], mode='bilinear', align_corners=True).squeeze(1)
+
+    min_vs_m = vs_m.min(dim=-1, keepdim=True).values.min(dim=-2, keepdim=True).values
+    max_vs_m = vs_m.max(dim=-1, keepdim=True).values.max(dim=-2, keepdim=True).values
+    return (vs_m - min_vs_m) / (max_vs_m - min_vs_m + eps)
+
+
+def rgb2xyz(x: torch.Tensor) -> torch.Tensor:
+    """
```

Comment:
Use raw docstring format 😄 
```suggestion
    r"""
```